src/prognostic_equations/implicit/auto_dense_jacobian.jl

```diff
@@ -0,0 +1,300 @@
+import ForwardDiff
+import LinearAlgebra: diagind
+
+import ClimaComms
+import ClimaCore.DataLayouts: MArray
+
+"""
+    AutoDenseJacobian([max_simultaneous_derivatives])
```

Comment:
This is a clear documentation explaining the logic behind the algorithm well and containing useful references. I was able to understand the algorithm and roughly what the code is meant to do based on the documentation.